scripts/ubuntu-bartender/aws-provider

```diff
@@ -89,15 +91,27 @@ function build-provider-create {
 }
 
 function build-provider-upload {
-  scp -qo StrictHostKeyChecking=no "$2" "ubuntu@$ip_addr:$3"
+  set-key-opt
+
+  scp ${KEY_OPT} -qo StrictHostKeyChecking=no "$2" "ubuntu@$ip_addr:$3"
 }
 
 function build-provider-download {
-  scp -qo StrictHostKeyChecking=no "ubuntu@$ip_addr:$2" "$3"
+  set-key-opt
+
+  scp ${KEY_OPT} -qo StrictHostKeyChecking=no "ubuntu@$ip_addr:$2" "$3"
 }
 
 function build-provider-run {
+  set-key-opt
+
   args=$(echo -- $@ | sed 's/.*--//')
-  ssh -qo StrictHostKeyChecking=no "ubuntu@$ip_addr" -- $args
+  ssh ${KEY_OPT} -qo StrictHostKeyChecking=no "ubuntu@$ip_addr" -- $args
 }
 
+function set-key-opt() {
+  export KEY_OPT=''
+  if [ -f ~/.ssh/${AWS_KEYPAIR_NAME}.pem ]; then
+    KEY_OPT="-i ~/.ssh/${AWS_KEYPAIR_NAME}.pem"
+  fi
+}
```

Comment:
Since bash is dynamically scoped, you can just call the contents of `set-key-opt` here, and the values of KEY_OPT will be available to all of the build-provider-* functions. No need to call it multiple times :-)

Or, better yet, have set-key-opt be a function that _echo_ s `KEY_OPT`, then you can pull it in directly like `ssh $(key-opt) -qo ...` . That way you avoid the messiness of dynamic scoping.

Reply:
Ok, I've pushed an additional commit. I've left the `export KEY_OPT` statement in, as it could be useful to see the variable's value in the environment during debugging.